### PR TITLE
Cataloger entries scan iterator

### PR DESCRIPTION
**Pull request by tzahij**

### CODE REVIEW

File: catalog/db_branch_reader.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+	_ = s
+	continueationQuery := baseSelect(r.branchID, r.commitID)
+	continueationQuery = continueationQuery.
+		Where("Path > ?", *lastRow.Path).
+		Limit(uint64(r.bufSize))
+	s = sq.DebugSqlizer(continueationQuery)
+	_ = s
+	// move rows of last Path to beginning of buffer
+	tempBuf := make([]*entryPK, 0, r.bufSize+len(r.buf)*2)
+	tempBuf = append(tempBuf, r.buf...)
```
Comment (nopcoder) [inline catalog/db_branch_reader.go:214]: use 'copy'

File: catalog/db_branch_reader.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	} else if commitID > 0 {
+		q = q.Where("min_commit between 1 and ?", commitID)
+	}
+	return q
+}
+
+func (r *singleBranchReader) extendBuf() error {
+	l := len(r.buf)
+	if l == 0 {
+		panic("in extendBuf with empty buffer!!!!!")
```
Comment (nopcoder) [inline catalog/db_branch_reader.go:199]: don't call panic

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: catalog/db_branch_reader.go [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+}
+type singleBranchReader struct {
+	tx        db.Tx
+	branchID  int64
+	buf       []*entryPK
+	bufSize   int
+	EOF       bool
+	after     string
+	commitID  CommitID
+	firstTime bool
```
[the remark this answers is not in the document]
Comment (tzahij) [reply catalog/db_branch_reader.go:23]: There may be cases where buf==nil at end of reading. I Think the "firstTime" approach is clearer, and worth the few extra lines

File: catalog/db_branch_reader.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+		tx:        tx,
+		branchID:  branchID,
+		bufSize:   bufSize,
+		after:     after,
+		commitID:  commitID,
+		firstTime: true,
+	}
+}
+
+func NewLineageReader(tx db.Tx, branchID int64, commitID CommitID, bufSize, limit int, after string) *lineageReader {
```
[the remark this answers is not in the document]
Comment (tzahij) [reply catalog/db_branch_reader.go:49]: will rename

File: catalog/db_branch_reader.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+		return ret
+	}
+	//todo: rethink condition
+	if ret.MaxCommit >= lineageCommitID {
+		ret.MaxCommit = MaxCommitID
+	}
+	return ret
+}
+
+func baseSelect(branchID int64, commitID CommitID) sq.SelectBuilder {
```
[the remark this answers is not in the document]
Comment (tzahij) [reply catalog/db_branch_reader.go:183]: branchSelectWithCommitID

File: catalog/db_branch_reader.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	if l == 0 {
+		panic("in extendBuf with empty buffer!!!!!")
+	}
+	lastRow := r.buf[l-1]
+	completionQuery := baseSelect(r.branchID, r.commitID)
+	completionQuery = completionQuery.Where("Path = ? and min_commit < ?", *lastRow.Path, lastRow.MinCommit)
+	s := sq.DebugSqlizer(completionQuery)
+	_ = s
+	continueationQuery := baseSelect(r.branchID, r.commitID)
+	continueationQuery = continueationQuery.
```
[the remark this answers is not in the document]
Comment (tzahij) [reply catalog/db_branch_reader.go:207]: done

File: catalog/db_branch_reader.go (hunk 1 above)
Comment (tzahij) [reply catalog/db_branch_reader.go:214]: Will change all of this once I change min-commit to MaxCommitID

File: catalog/db_branch_reader.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+func union(compleateCurrntPath, continueationQuery sq.SelectBuilder) sq.SelectBuilder {
+	unionQuery := compleateCurrntPath.
+		Prefix("(").
+		SuffixExpr(sq.ConcatExpr(")\n UNION ALL \n(", continueationQuery, ")"))
+	s := sq.DebugSqlizer(unionQuery)
+	_ = s
+	return unionQuery
+}
+
+func fillBuf(tx db.Tx, q sq.SelectBuilder, buf *[]*entryPK) error {
```
[the remark this answers is not in the document]
Comment (tzahij) [reply catalog/db_branch_reader.go:231]: Changed name, but prefer to keep this function as it is called from two places, and it saves a few lines. It is the only database call ion this code

File: catalog/db_branch_reader.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	sql, args, err := q.PlaceholderFormat(sq.Dollar).ToSql()
+	if err != nil {
+		return err
+	}
+	err = tx.Select(buf, sql, args...)
```
[the remark this answers is not in the document]
Comment (tzahij) [reply catalog/db_branch_reader.go:236]: change to return err. do not like to return directly, as sometimes it makes debugging harder

File: catalog/db_branch_reader.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+		branchID:  branchID,
+		bufSize:   bufSize,
+		after:     after,
+		commitID:  commitID,
+		firstTime: true,
+	}
+}
+
+func NewLineageReader(tx db.Tx, branchID int64, commitID CommitID, bufSize, limit int, after string) *lineageReader {
+	// limit <= 0 means there is no limit to number of returned rows
```
[the remark this answers is not in the document]
Comment (tzahij) [reply catalog/db_branch_reader.go:50]: has more does not work at this level. the levels above have already added 1 to understand the "has more"

**Review by nopcoder — APPROVED** [review]

**Pull request merged**